Font Library: Add all REST API endpoints.

youknowriad opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/59166

This builds on top of the initial [Font Library API PR ](https://github.com/WordPress/wordpress-develop/pull/6027) and adds the necessary REST API controllers to make the Font Library work properly.

### Testing instructions

- You can add the following piece of code anywhere

```
function gutenberg_register_font_collections() {
	wp_register_font_collection( 'google-fonts', 'https://raw.githubusercontent.com/WordPress/google-fonts-to-wordpress-collection/01aa57731575bd13f9db8d86ab80a2d74e28a1ac/releases/gutenberg-17.6/collections/google-fonts-with-preview.json' );
}
add_action( 'init', 'gutenberg_register_font_collections' );
```

- Go to the site editor
- Open the global styles panel
- Open typography
- Click on the "Aa" icon to open the font library
- On the third panel, you should be able to enable google fonts and use them properly.

getdave committed:
Font Library: Add all REST API endpoints.

youknowriad commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-collections-controller.php:
```diff
@@ -0,0 +1,329 @@
+ * @since      6.5.0
+ */
+
+if ( ! class_exists( 'WP_REST_Font_Collections_Controller' ) ) {
```
I need to remove these.

youknowriad commented:
Working on fixing the unit tests.

swissspidy commented on src/wp-includes/post.php:
```diff
@@ -588,8 +588,12 @@ function create_initial_post_types() {
+			'show_in_rest'                   => true,
+			'rest_base'                      => 'font-families',
+			'rest_controller_class'          => 'WP_REST_Font_Families_Controller',
+			// Disable autosave endpoints for font families.
+			'autosave_rest_controller_class' => 'stdClass',
```
**Aside:**

We should really find a better way to avoid hacks like this. I don't think this is a clean way at all, and devs will just copy this sooner or later.

If workarounds like this are added/needed in GB it's a sign that we should improve core, not just bring the workaround into core.

Maybe it could be handled via post type supports, i.e. `add_post_type_support( 'autosave' )` / `remove_post_type_support( 'autosave' )`. I suppose https://core.trac.wordpress.org/ticket/41172 could be used to track this, maybe I can take a stab at it.

youknowriad committed:
Fix unit tests

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-collections-controller.php:
```diff
@@ -0,0 +1,326 @@
+		$slug       = $request->get_param( 'slug' );
+		$collection = WP_Font_Library::get_instance()->get_font_collection( $slug );
+
+		// If the collection doesn't exist returns a 404.
+		if ( is_wp_error( $collection ) ) {
+			$collection->add_data( array( 'status' => 404 ) );
+			return $collection;
+		}
```
Odd that `get_font_collection()` returns a `WP_Error` and not just `false`. Especially since the method here just assumes that any `WP_Error` means a 404.

I think it would be cleaner for `get_font_collection()` to return `false` or perhaps `null`, and then this could here would just do

```suggestion
		if ( ! $collection ) {
			return new WP_Error( 'font_collection_not_found', __( 'Font collection not found.' ), array( 'status' => 404 ) );
		}
```

Aside: the comment is a bit superfluous, the code is self-explanatory here

matiasbenedetto replied:
Good catch, thanks. I addressed your feedback in this Gutenberg PR: https://github.com/WordPress/gutenberg/pull/58735. Once merged, I'll commit that to this PR.

swissspidy commented:
@youknowriad In #6027 it was mentioned that `\WP_Font_Utils::get_allowed_font_mime_types()` (which so far is only used in tests) is needed for the REST API. But I don't see it being used here at all. What am I missing?

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-families-controller.php:
```diff
@@ -0,0 +1,554 @@
+			return new WP_Error(
+				'rest_trash_not_supported',
+				/* translators: %s: force=true */
+				sprintf( __( "Font faces do not support trashing. Set '%s' to delete." ), 'force=true' ),
```
In other places we use double quotes around the placeholder, not single quotes.

```suggestion
				sprintf( __( 'Font faces do not support trashing. Set "%s" to delete.' ), 'force=true' ),
```

matiasbenedetto replied:
I'm addressing this in this Gutenberg PR: https://github.com/WordPress/gutenberg/pull/58736. Once merged, I'll commit to this PR.

swissspidy replied:
Why do we have this constant back and forth? 🤔

matiasbenedetto replied:
I think it's useful to merge the changes in Gutenberg first to allow more granular control of them and to use the code from Gutenberg as a reference and source of truth in case of code conflicts. IMO, it is easier to handle code changes moving code always only in one direction: GB --> Core instead of GB <--->  Core.

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-families-controller.php:
```diff
@@ -0,0 +1,554 @@
+		$query_params = parent::get_collection_params();
+
+		// Remove unneeded params.
+		unset( $query_params['after'] );
+		unset( $query_params['modified_after'] );
+		unset( $query_params['before'] );
+		unset( $query_params['modified_before'] );
+		unset( $query_params['search'] );
+		unset( $query_params['search_columns'] );
+		unset( $query_params['status'] );
```
Nit: you can just unset these in one call:

`unset( $query_params['after'], $query_params['modified_after'], ... )`

matiasbenedetto replied:
I'm addressing this in this Gutenberg PR: https://github.com/WordPress/gutenberg/pull/58736. Once merged, I'll commit to this PR.

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-families-controller.php:
```diff
@@ -0,0 +1,554 @@
+			)
+		);
+
+		return $query->get_posts();
```
Calling `new WP_Query()` already calls `get_get_posts()`, so this extra `get_posts()` call is superfluous as it just queries everything twice.

```suggestion
		return $query->posts;
```

matiasbenedetto replied:
I'm addressing this in this Gutenberg PR: https://github.com/WordPress/gutenberg/pull/58736. Once merged, I'll commit to this PR.

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-families-controller.php:
```diff
@@ -0,0 +1,554 @@
+		foreach ( $font_face_ids as $font_face_id ) {
+			$links[] = array(
+				'embeddable' => true,
+				'href'       => rest_url( $this->namespace . '/' . $this->rest_base . '/' . $font_family_id . '/font-faces/' . $font_face_id ),
```
Nit: in other REST controllers we usually use `sprintf` for assembling REST URLs like this.

matiasbenedetto replied:
I'm addressing this in this Gutenberg PR: https://github.com/WordPress/gutenberg/pull/58736. Once merged, I'll commit to this PR.

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-faces-controller.php:
```diff
@@ -0,0 +1,941 @@
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable -- required by parent class
```
FWIW all these PHPCS warnings are not emitted in core code, so might as well remove the comments

matiasbenedetto replied:
Thanks for pointing that out. We should probably have the same linting rules in both Gutenberg and Core.
Anyway that's out of the scope of this PR so I removed the comments in this commit: https://github.com/WordPress/wordpress-develop/pull/6034/commits/3ddf0235098ecdb1fec572649ff360b295b7c604

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-faces-controller.php:
```diff
@@ -0,0 +1,941 @@
+		$query_params = parent::get_collection_params();
+
+		// Remove unneeded params.
+		unset( $query_params['after'] );
+		unset( $query_params['modified_after'] );
+		unset( $query_params['before'] );
+		unset( $query_params['modified_before'] );
+		unset( $query_params['search'] );
+		unset( $query_params['search_columns'] );
+		unset( $query_params['slug'] );
+		unset( $query_params['status'] );
```
Same here: use one `unset()` call for all of these

matiasbenedetto replied:
I'm addressing this in this Gutenberg PR: https://github.com/WordPress/gutenberg/pull/58736. Once merged, I'll commit to this PR.

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-families-controller.php:
```diff
@@ -0,0 +1,554 @@
+				'update_post_term_cache' => false,
+			)
+		);
+		if ( ! empty( $query->get_posts() ) ) {
```
Same here, the `WP_Query` constructor already calls `get_posts()`. This just causes duplicate work.

```suggestion
		if ( ! empty( $query->posts ) ) {
```

matiasbenedetto replied:
I'm addressing this in this Gutenberg PR: https://github.com/WordPress/gutenberg/pull/58736. Once merged, I'll commit to this PR.

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-faces-controller.php:
```diff
@@ -0,0 +1,941 @@
+				'update_post_term_cache' => false,
+			)
+		);
+		if ( ! empty( $query->get_posts() ) ) {
```
Same here, the constructor already calls `get_posts()`.

```suggestion
		if ( ! empty( $query->posts ) ) {
```

matiasbenedetto replied:
I'm addressing this in this Gutenberg PR: https://github.com/WordPress/gutenberg/pull/58736. Once merged, I'll commit to this PR.

youknowriad commented:
@swissspidy Yes, that function is used here https://github.com/WordPress/wordpress-develop/blob/2eb4cc3650ff80208bb2e234b01920cf80dc4058/src/wp-includes/rest-api/endpoints/class-wp-rest-font-faces-controller.php#L848-L872

youknowriad commented:
Anyone knows how to regenerate the `wp-api-generated.js` fixture?

ocean90 commented:
> Anyone knows how to regenerate the `wp-api-generated.js` fixture?

Try running:
```
npm run test:php -- --filter test_build_wp_api_client_fixtures
```

or:

```
npm run test:php -- --filter WP_Test_REST_Schema_Initialization
```

youknowriad committed:
Regenerate api fixtures

youknowriad committed:
Generate the right fixtures

youknowriad commented:
@ocean90 Both of these options return a different version of that file 🤷 Not sure what I'm doing wrong.

mcsf commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-families-controller.php:
```diff
@@ -0,0 +1,554 @@
+		$required = $schema['required'];
+
+		if ( isset( $request['id'] ) ) {
+			// Allow sending individual properties if we are updating an existing font family.
+			unset( $schema['required'] );
```
This value is unset here, but it has already been captured into `$required`. Later on, at the end of the method, we iterate over it using `foreach ( $required as $key )`. Is this all intended, or is there a bug hidden here? For example, was the intention actually something like:

```diff
		// Check that none of the required settings are empty values.
-		foreach ( $required as $key ) {
+		if ( isset( $schema['required'] ) ) {
+			foreach ( $required as $key ) {
```

creativecoder replied:
To walk through what's going on here, currently:

If a new font family is being created

- `$schema['required']` stays set
- `rest_validate_value_from_schema` validates that those required properties are set (but does not error when they are empty)
- `foreach ( $required as $key ) { ...` checks that required properties are not empty strings

If an existing font family is being updated ( `$request['id']` is set)

- `$schema['required']` is unset, because individual properties (except for slug) can be updated, so no properties are required
- `rest_validate_value_from_schema` validates that properties in the request match the schema (but does not check for required properties)
- `foreach ( $required as $key ) { ...` checks that any schema required properties that are set in the request are not empty

> Is this all intended, or is there a bug hidden here?

It's intended. There are a number of unit tests covering font family creation and updating, which should catch any logic errors here (See `Tests_REST_WpRestFontFamiliesController::test_create_item_invalid_settings` and `Tests_REST_WpRestFontFamiliesController::test_update_item_empty_settings`).

mcsf replied:
Thanks for the comprehensive answer

mcsf commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-families-controller.php:
```diff
@@ -0,0 +1,554 @@
+		}
+
+		if ( rest_is_field_included( 'theme_json_version', $fields ) ) {
+			$data['theme_json_version'] = 2;
```
Is this the right place for a magic number like `2`?

matiasbenedetto replied:
I think it makes sense to use `WP_Theme_JSON::LATEST_SCHEMA` here. Added in this commit https://github.com/WordPress/gutenberg/pull/58736/commits/99c98b0a6d653a7f1ccbd42e8021d045afc21ae1 from this PR: https://github.com/WordPress/gutenberg/pull/58736. I'll commit here once that GB PR is merged.

creativecoder replied:
We could use `WP_Theme_JSON::LATEST_SCHEMA` here.

But let's say there's a theme.json update and the structure of font families and/or font faces in theme.json are changed. The controllers would need to be updated to support that new structure.

Since we don't know exactly what changes a new theme.json version might bring, I think it's better to manually update the versions that the endpoints accept, after support for a new version is verified.

In the phpunit tests we are [checking to see if the theme_json_version here matches](https://github.com/WordPress/wordpress-develop/pull/6034/files#diff-3075dbacbce6e5a60cdfdd338cea96b6ed5889569c397d7832615c1a43811fafR1001) `WP_Theme_JSON::LATEST_SCHEMA` -- when the latest version of theme.json changes, that would signal an update is needed to these controllers to support the latest version.

mcsf replied:
> But let's say there's a theme.json update and the structure of font families and/or font faces in theme.json are changed. The controllers would need to be updated to support that new structure.

Fair enough! Can we then capture the `2` in a private variable? That way the constant can be documented and link back to the phpunit tests in question.

creativecoder replied:
Yes, I think that's a good idea! Updating in Gutenberg to add a constant for each controller, and add an explicit test to check against the latest theme.json version, see https://github.com/WordPress/gutenberg/pull/58736

creativecoder replied:
This PR has been updated to use `const LATEST_THEME_JSON_VERSION_SUPPORTED = 2;` in font family/face controllers and in related tests.

mcsf commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-families-controller.php:
```diff
@@ -0,0 +1,554 @@
+				'theme_json_version'   => array(
+					'description' => __( 'Version of the theme.json schema used for the typography settings.' ),
+					'type'        => 'integer',
+					'default'     => 2,
+					'minimum'     => 2,
+					'maximum'     => 2,
```
Same question about the magic number. What happens when theme.json gets a new version? Who will update this array?

matiasbenedetto replied:
I think it makes sense to use `WP_Theme_JSON::LATEST_SCHEMA` here. Added in this commit https://github.com/WordPress/gutenberg/commit/99c98b0a6d653a7f1ccbd42e8021d045afc21ae1 from this PR: https://github.com/WordPress/gutenberg/pull/58736. I'll commit here once that GB PR is merged.

creativecoder replied:
[See my comment above](https://github.com/WordPress/wordpress-develop/pull/6034/files#r1480150055), considering how an update process might work.

mcsf reviewed (commented):
Looking pretty good.

matiasbenedetto committed:
remove PHPCS comments not needed in core

requested in https://github.com/WordPress/wordpress-develop/pull/6034#discussion_r1479458090

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-faces-controller.php:
```diff
@@ -0,0 +1,941 @@
+	 * @since 6.5.0
+	 *
+	 * @param string          $value   Encoded JSON string of font face settings.
+	 * @param WP_REST_Request $request Request object.
```
There is no `$request` param in this method

```suggestion
```

creativecoder replied:
PR now updated to remove the extra `@param` and improve the `@return` description.

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-faces-controller.php:
```diff
@@ -0,0 +1,941 @@
+	 * @since 6.5.0
+	 *
+	 * @param array $file Single file item from $_FILES.
+	 * @return array Array containing uploaded file attributes on success, or error on failure.
```
Since this method can also return a `WP_Error`, this needs to be adjusted accordingly.

```suggestion
	 * @return array|WP_Error Array containing uploaded file attributes on success, or error on failure.
```

creativecoder replied:
Updated in https://github.com/WordPress/wordpress-develop/pull/6034/commits/a0836f0f6b4fffc3259a2711f2ea43b02823a6e2

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-faces-controller.php:
```diff
@@ -0,0 +1,941 @@
+	 * @since 6.5.0
+	 *
+	 * @param WP_REST_Request $request Request object.
+	 * @return stdClass|WP_Error Post object or WP_Error.
```
This actually never returns an error.

```suggestion
	 * @return stdClass Post object.
```

creativecoder replied:
Updated in https://github.com/WordPress/wordpress-develop/pull/6034/commits/a0836f0f6b4fffc3259a2711f2ea43b02823a6e2

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-collections-controller.php:
```diff
@@ -0,0 +1,326 @@
+	* @param WP_REST_Request    $request    Request object.
+	* @return WP_REST_Response Response object.
+	*/
+	public function prepare_item_for_response( $collection, $request ) {
```
To be in line with other REST controllers, this parameter's name should be `$item`.

The other `$item` inside the method below should be renamed to `$data`.

creativecoder replied:
Good point, as the name `$collection` could be especially confusing here, given it's a single "item". Updated in https://github.com/WordPress/wordpress-develop/pull/6034/commits/a0836f0f6b4fffc3259a2711f2ea43b02823a6e2

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-collections-controller.php:
```diff
@@ -0,0 +1,326 @@
+	*
+	* @param WP_Font_Collection $collection Collection object.
+	* @param WP_REST_Request    $request    Request object.
+	* @return WP_REST_Response Response object.
```
```suggestion
	* @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
```

Because it can return an error

creativecoder replied:
Updated in https://github.com/WordPress/wordpress-develop/pull/6034/commits/a0836f0f6b4fffc3259a2711f2ea43b02823a6e2

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-families-controller.php:
```diff
@@ -0,0 +1,554 @@
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) { 
```
```suggestion
	public function get_items_permissions_check( $request ) {
```

Whitespace at end of line is not allowed

creativecoder replied:
Updated in https://github.com/WordPress/wordpress-develop/pull/6034/commits/a0836f0f6b4fffc3259a2711f2ea43b02823a6e2

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-families-controller.php:
```diff
@@ -0,0 +1,554 @@
+	 * @since 6.5.0
+	 *
+	 * @param string          $value   Encoded JSON string of font family settings.
+	 * @param WP_REST_Request $request Request object.
```
There is no `$request` parameter here

```suggestion
```

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-families-controller.php:
```diff
@@ -0,0 +1,554 @@
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string          $value   Encoded JSON string of font family settings.
+	 * @param WP_REST_Request $request Request object.
```
There is no `$request` parameter here

```suggestion
	 * @param string $value Encoded JSON string of font family settings.
```

creativecoder replied:
Updated in https://github.com/WordPress/wordpress-develop/pull/6034/commits/a0836f0f6b4fffc3259a2711f2ea43b02823a6e2

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-families-controller.php:
```diff
@@ -0,0 +1,554 @@
+	 *
+	 * @param string          $value   Encoded JSON string of font family settings.
+	 * @param WP_REST_Request $request Request object.
+	 * @return array                   Decoded array font family settings.
```
```suggestion
	 * @return array Decoded array font family settings.
```

The return description must not be aligned with the param description

creativecoder replied:
Updated in https://github.com/WordPress/wordpress-develop/pull/6034/commits/a0836f0f6b4fffc3259a2711f2ea43b02823a6e2

aaronjorbin commented on tests/phpunit/tests/rest-api/rest-schema-setup.php:
```diff
@@ -189,6 +189,13 @@ public function test_expected_routes_in_schema() {
+			'/wp/v2/font-families/(?P<font_family_id>[\d]+)/font-faces',
+			'/wp/v2/font-families/(?P<font_family_id>[\d]+)/font-faces/(?P<id>[\d]+)',
+			'/wp/v2/font-families/(?P<id>[\d]+)',
+
```
Unnecessary empty line
```suggestion
```

creativecoder replied:
Updated in https://github.com/WordPress/wordpress-develop/pull/6034/commits/2cf254702e0747b2701b1e52d789ec553e4ff96d

creativecoder committed:
Updates from gutenberg #58752, #58735, and #58736 to address feedback comments

creativecoder committed:
Remove extra empty line in rest-schema-setup.php

creativecoder committed:
Attempt to fix unit tests by commiting updates to wp-api-generated.js from test run

youknowriad commented:
Looks like all the feedback have been addressed here. I'll be committing this shortly if there's no other blocker.

youknowriad commented:
Commit https://core.trac.wordpress.org/changeset/57548

lutzmex commented:
What steps are taken to ensure that the Font Library seamlessly integrates with various WordPress themes, allowing users to activate and deactivate fonts independently for each theme ?
<a href="https://dinetuatuwhoo.tk"> </a><a href="https://dickblenag.tk"> </a><a href="https://ceoplasthonnilanwho.tk"> </a><a href="https://apimslufli.tk"> </a><a href="https://enycuvceled.tk"> </a><a href="https://mowuwyjonofo.tk"> </a><a href="https://daybedfrapetdoso.tk"> </a><a href="https://spormorandpor.tk"> </a><a href="https://turbuitragavflatam.tk"> </a><a href="https://inaltnicme.tk"> </a><a href="https://leraphythepapa.tk"> </a><a href="https://citretalingros.tk"> </a><a href="https://resstragabli.tk"> </a><a href="https://mimapitge.tk"> </a><a href="https://quetewestboncunext.tk"> </a><a href="https://hardpishuntjaztiwong.tk"> </a><a href="https://sichttraftiosio.tk"> </a><a href="https://plesbubbversnakawigg.tk"> </a><a href="https://nnelonlapuscli.tk"> </a><a href="https://grahitebmiron.tk"> </a><a href="https://swholnessplanunanslum.tk"> </a><a href="https://palisutela.tk"> </a><a href="https://paisnufancami.cf"> </a><a href="https://dripizinma.cf"> </a><a href="https://paigrahec.cf"> </a><a href="https://stuntersosafafal.cf"> </a><a href="https://florbulknamat.tk"> </a><a href="https://tretexenexenin.tk"> </a><a href="https://klebtalsiboto.tk"> </a><a href="https://enfifernbungvas.tk"> </a><a href="https://atesurpupat.tk"> </a><a href="https://contmirama.tk"> </a><a href="https://laudisfactmeti.tk"> </a><a href="https://chondconna.tk"> </a><a href="https://goffmerpenoradco.tk"> </a><a href="https://indaloticco.tk"> </a><a href="https://tiutravkyokoma.tk"> </a><a href="https://palpaddkinisort.tk"> </a><a href="https://keypleserar.tk"> </a><a href="https://puecryssechighfrom.tk"> </a><a href="https://ulunaritxyp.gq"> </a><a href="https://sollmarctesmopha.gq"> </a><a href="https://tiasurphokatqui.gq"> </a><a href="https://tiecereti.cf"> </a><a href="https://bronsouthchotedtai.cf"> </a><a href="https://buckbamentwerbrive.cf"> </a><a href="https://manorecepe.cf"> </a><a href="https://debifilicsa.cf"> </a><a href="https://icathcellomo.tk"> </a><a href="https://dwinenclan.tk"> </a><a href="https://pratanin.tk"> </a><a href="https://reilassoco.tk"> </a><a href="https://lanttowa.tk"> </a><a href="https://ucolapezon.tk"> </a><a href="https://jasubssusubull.tk"> </a><a href="https://raimenbodefsacyc.tk"> </a><a href="https://rucktrofarexliswa.tk"> </a><a href="https://ilessanmi.tk"> </a><a href="https://paiburgthing.tk"> </a><a href="https://apcanpa.tk"> </a><a href="https://diabroxlighmen.tk"> </a><a href="https://subbusas.tk"> </a><a href="https://javipiwotywu.tk"> </a><a href="https://wieryoglycachexes.tk"> </a><a href="https://berpgomi.tk"> </a><a href="https://alglitar.tk"> </a><a href="https://ttemalba.tk"> </a><a href="https://yqihikutej.tk"> </a><a href="https://monbacolcompcelta.tk"> </a><a href="https://schizinseavo.tk"> </a><a href="https://buiprovdownnasdexi.tk"> </a><a href="https://fractentdaboo.tk"> </a><a href="https://partlafomenpo.tk"> </a><a href="https://emcepmebelgeber.tk"> </a><a href="https://xperunisspinan.tk"> </a><a href="https://funccalfonctinke.tk"> </a><a href="https://ampradarygeder.tk"> </a><a href="https://pieproofviscagua.tk"> </a><a href="https://olehthrusobinov.tk"> </a><a href="https://phomosufivepa.tk"> </a><a href="https://tidatelecanest.tk"> </a><a href="https://gavuldeno.tk"> </a><a href="https://halnavousaltdis.tk"> </a><a href="https://lobeldisscago.tk"> </a><a href="https://mantatbschumined.tk"> </a><a href="https://fonthoupostrapan.tk"> </a><a href="https://dispdara.tk"> </a><a href="https://nocumrare.tk"> </a><a href="https://wezusotacom.tk"> </a><a href="https://nortiju.tk"> </a><a href="https://megetudype.tk"> </a><a href="https://avpamaluhedor.tk"> </a><a href="https://bundcontne.tk"> </a><a href="https://licfitun.tk"> </a><a href="https://cesworklopho.tk"> </a><a href="https://ticppretarisorres.tk"> </a><a href="https://witzmaltvestgreg.tk"> </a><a href="https://ulergrapherrared.tk"> </a><a href="https://urbocana.tk"> </a><a href="https://trifolinco.tk"> </a><a href="https://traprailiber.tk"> </a><a href="https://kalsopaket.tk"> </a><a href="https://cetocoolte.tk"> </a><a href="https://vitacmapa.tk"> </a><a href="https://orelunnimon.tk"> </a><a href="https://plowrysdekaptoulu.tk"> </a><a href="https://poilighsigehyfos.tk"> </a><a href="https://inalmacsarusis.tk"> </a><a href="https://magziebreathim.tk"> </a><a href="https://landsusvprecintermai.tk"> </a><a href="https://yplirima.tk"> </a><a href="https://lwortheardempproc.tk"> </a><a href="https://banraldtrad.tk"> </a><a href="https://mcenisraszaimin.tk"> </a><a href="https://joyfiticsdegeva.tk"> </a><a href="https://plotsakami.tk"> </a><a href="https://freewinmerzirohy.tk"> </a><a href="https://noespastyma.tk"> </a><a href="https://stylnaltciconla.tk"> </a><a href="https://profeqiladit.tk"> </a><a href="https://supsegatluabal.tk"> </a><a href="https://ewexafra.tk"> </a><a href="https://handsunriggvervarid.tk"> </a><a href="https://brabadclimer.tk"> </a><a href="https://nimephuddcudeles.tk"> </a><a href="https://roelotanciesbadup.tk"> </a><a href="https://reofreedarpi.tk"> </a><a href="https://lonnebagpaver.tk"> </a><a href="https://issarusttisiti.tk"> </a><a href="https://pletaltica.tk"> </a><a href="https://inrekadasel.tk"> </a><a href="https://unperichttima.tk"> </a><a href="https://manlocenpuloosub.tk"> </a><a href="https://thaetare.tk"> </a><a href="https://cordepearati.tk"> </a><a href="https://decuramicur.tk"> </a><a href="https://esperedubigs.tk"> </a><a href="https://bynpomusipas.tk"> </a><a href="https://thospeineodeton.tk"> </a><a href="https://bronexhorbese.tk"> </a><a href="https://pisidar.tk"> </a><a href="https://sookdlanpinagmela.tk"> </a><a href="https://meghderlibudddis.tk"> </a><a href="https://tictaincarneomag.tk"> </a><a href="https://tradenmy.tk"> </a><a href="https://camptophyve.tk"> </a><a href="https://hoalilracon.tk"> </a><a href="https://subseibuzlingganar.tk"> </a><a href="https://vessixthvertfor.gq"> </a><a href="https://piedripchanluver.gq"> </a><a href="https://apibiz.gq"> </a><a href="https://partderlatertu.gq"> </a><a href="https://setsporttenpoenia.gq"> </a><a href="https://tenshighrasy.gq"> </a><a href="https://stilhorcampnetroll.gq"> </a><a href="https://llemsecmudary.cf"> </a><a href="https://dritcalma.cf"> </a><a href="https://mighnasle.cf"> </a><a href="https://mierenrakuti.cf"> </a><a href="https://persundtranabedpai.cf"> </a><a href="https://buywinsswin.cf"> </a><a href="https://ocesanbitu.cf"> </a><a href="https://haiciruss.cf"> </a><a href="https://huispilelinemle.cf"> </a><a href="https://kachahirodens.cf"> </a><a href="https://layhona.cf"> </a><a href="https://vetowncolbert.cf"> </a><a href="https://nieprocempo.cf"> </a><a href="https://tingnoricide.cf"> </a><a href="https://ranguilystegubley.cf"> </a><a href="https://ennelnate.tk"> </a><a href="https://spacinaflorama.tk"> </a><a href="https://vipynikedpaytroub.tk"> </a><a href="https://raitigadrada.tk"> </a><a href="https://raponolsereren.tk"> </a><a href="https://franyctepiglo.tk"> </a><a href="https://frusatspacsuppcon.tk"> </a><a href="https://inrogchondvuldownver.tk"> </a><a href="https://builongcalleftvercau.tk"> </a><a href="https://softlinssivemil.tk"> </a><a href="https://unrirejecforc.tk"> </a><a href="https://ingolemeatre.tk"> </a><a href="https://riaviegucinobou.tk"> </a><a href="https://voicountporegerca.tk"> </a><a href="https://aptuydolnalote.tk"> </a><a href="https://ferlimamithylttung.tk"> </a><a href="https://blensorchocyma.tk"> </a><a href="https://iktancadici.tk"> </a><a href="https://phalfobobgere.tk"> </a><a href="https://snowhrinlafestwheewa.tk"> </a><a href="https://trophambsedebtrip.tk"> </a><a href="https://samconsderlyro.tk"> </a><a href="https://rizdoubtmo.tk"> </a><a href="https://tropnirimepho.tk"> </a><a href="https://amnalabatt.tk"> </a><a href="https://mestiticomdia.tk"> </a><a href="https://kairalode.tk"> </a><a href="https://dressethukspangcualbao.tk"> </a><a href="https://tarjahy.tk"> </a><a href="https://svolatnewgecon.tk"> </a><a href="https://hungtyptorchsubpo.tk"> </a><a href="https://lensloncomalphol.tk"> </a><a href="https://thoughgetedichatmo.tk"> </a><a href="https://pirehy.tk"> </a><a href="https://didetutejosbe.tk"> </a><a href="https://fitzconricarwaru.tk"> </a><a href="https://coakitatacimor.tk"> </a><a href="https://ogtimamoramig.tk"> </a><a href="https://gintkurzwhomanlenum.tk"> </a><a href="https://seiloncoorehun.tk"> </a><a href="https://sevesouthlipas.tk"> </a><a href="https://curdperretuff.tk"> </a><a href="https://elendiabalucgee.tk"> </a><a href="https://destjopazeana.tk"> </a><a href="https://manbucklesgmake.tk"> </a><a href="https://precicurturisi.tk"> </a><a href="https://cauclathutaldab.tk"> </a><a href="https://taijourhaacribon.tk"> </a>
